data/src/main/java/me/androidbox/data/remote/authentication/AuthenticationRepositoryImp.kt

```diff
@@ -0,0 +1,66 @@
+package me.androidbox.data.remote.authentication
+
+import kotlinx.coroutines.CancellationException
+import me.androidbox.data.remote.model.request.LoginRequestDto
+import me.androidbox.data.remote.model.request.RegistrationRequestDto
+import me.androidbox.data.remote.network.authentication.AuthenticationService
+import me.androidbox.domain.authentication.NetworkResponseState
+import me.androidbox.domain.authentication.model.LoginModel
+import me.androidbox.domain.authentication.remote.AuthenticationRepository
+import retrofit2.HttpException
+import javax.inject.Inject
+
+class AuthenticationRepositoryImp @Inject constructor(
+    private val authenticationService: AuthenticationService,
+) : AuthenticationRepository {
+
+    override suspend fun registerUser(fullName: String, email: String, password: String): NetworkResponseState<Unit> {
+        val registrationRequestDto = RegistrationRequestDto(
+            fullName = fullName,
+            email = email,
+            password = password
+        )
+
+        return try {
+            authenticationService.register(registrationRequestDto)
+            NetworkResponseState.Success(Unit)
+        }
+        catch(httpException: HttpException) {
+             NetworkResponseState.Failure(httpException)
+        }
+        catch (exception: Exception) {
+            if(exception is CancellationException) {
+                throw Exception()
+            }
+            NetworkResponseState.Failure(exception)
+        }
+    }
+
+    override suspend fun loginUser(email: String, password: String): NetworkResponseState<LoginModel> {
+        val loginRequestDto = LoginRequestDto(
+            email = email,
+            password = password
+        )
+
+        return try {
+            val loginDto = authenticationService.login(loginRequestDto)
+
+            val loginModel = LoginModel(
+                token = loginDto.token,
+                userId = loginDto.userId,
+                fullName = loginDto.fullName
+            )
+
+            return NetworkResponseState.Success(loginModel)
+        }
+        catch(httpException: HttpException) {
+            NetworkResponseState.Failure(httpException)
+        }
+        catch (exception: Exception) {
+            if(exception is CancellationException) {
+                throw Exception()
+            }
+            NetworkResponseState.Failure(exception)
+        }
```

Comment:
Since you are going to repeat this many times, you could have a function that does the following

Receive a lambda in the parameters, return a generic Result, and then it does the try catch, where in the try it calls the lambda and returns the success, and the catch returns the Result.failure(e) (and if it's cancellation exception, it throws it)

Reply:
@MKiperszmid 
I was thinking the same thing. I don't like to repeat myself either